kernel/src/log_segment.rs

```diff
@@ -112,13 +128,293 @@ impl LogSegment {
+                }
+                (None, Some(start_version), _) => {
+                    list_log_files_with_version(fs_client, &log_root, Some(start_version))?
+                }
+                _ => list_log_files_with_version(fs_client, &log_root, None)?,
+            };
+
+        if omit_checkpoint_files {
+            checkpoint_parts.clear();
+        }
+
+        // Commit file versions must satisfy the following:
+        // - Be greater than or equal to the start version
+        // - Be greater than the most recent checkpoint version if it exists
+        // - Be less than or equal to the end version.
+        if let Some(start_version) = start_version {
+            sorted_commit_files.retain(|log_path| log_path.version >= start_version);
+        }
+        if let Some(checkpoint_file) = checkpoint_parts.first() {
+            sorted_commit_files.retain(|log_path| checkpoint_file.version < log_path.version);
+        }
+        if let Some(end_version) = end_version {
+            sorted_commit_files.retain(|log_path| log_path.version <= end_version);
+        }
+
+        // After (possibly) omitting checkpoint files and filtering commits, we should have commits that are
+        // contiguous. In other words, there must be no gap between commit versions.
+        let ordered_commits = sorted_commit_files
+            .windows(2)
+            .all(|cfs| cfs[0].version + 1 == cfs[1].version);
+        if !ordered_commits {
+            return Err(Error::generic("Expected ordered contiguous commit files"));
+        }
+
+        // get the effective version from chosen files
+        let version_eff = sorted_commit_files
+            .last()
+            .or(checkpoint_parts.first())
+            .ok_or(Error::MissingVersion)? // TODO: A more descriptive error
+            .version;
+        if let Some(end_version) = end_version {
+            require!(
+                version_eff == end_version,
+                Error::MissingVersion // TODO more descriptive error
+            );
+        }
+
+        // We assume commit files are sorted in ascending order. If `sort_commit_files_ascending`
+        // is false, reverse to make it descending.
+        if !sort_commit_files_ascending {
+            sorted_commit_files.reverse();
+        }
+
+        Ok(LogSegment {
+            end_version: version_eff,
+            log_root,
+            commit_files: sorted_commit_files,
+            checkpoint_parts,
+        })
+    }
+}
+
+/// List all commit and checkpoint files with versions above the provided `version`. If successful, this returns
+/// a tuple `(sorted_commit_files_paths, checkpoint_parts): (Vec<ParsedLogPath>, Vec<ParsedLogPath>)`.
+/// The commit files are guaranteed to be sorted in ascending order by version. The elements of
+/// `checkpoint_parts` are all the parts of the same checkpoint. Checkpoint parts share the same
+/// version.
+fn list_log_files_with_version(
```

Comment:
This method was just moved (unchanged) from snapshot.rs?
Same for `list_log_files_with_checkpoint` below?

Reply:
`list_log_files_with_checkpoint` and `list_log_files_with_version` actually had much of the same code twice. I've merged the two so that you list and process files in one place.